reconcile/openshift_saas_deploy_trigger_base.py

```diff
@@ -0,0 +1,64 @@
+import logging
+
+from reconcile.jenkins_job_builder import get_openshift_saas_deploy_job_name
+
+
+def trigger(options):
+    """Trigger a deployment according to the specified pipelines provider
+
+    Args:
+        options (dict): A dictionary containing:
+            dry_run (bool): Is this a dry run
+            spec (dict): A trigger spec as created by saasherder
+            jenkins_map (dict): Instance names with JenkinsApi instances
+            already_triggered (list): A list of already triggered deployments
+            settings (dict): App-interface settings
+            state_update_method (function): A method to call to update state
+
+    Returns:
+        bool: True if there was an error, False otherwise
+    """
+    dry_run = options['dry_run']
+    spec = options['spec']
+    jenkins_map = options['jenkins_map']
+    already_triggered = options['already_triggered']
+    settings = options['settings']
+    state_update_method = options['state_update_method']
+
+    saas_file_name = spec['saas_file_name']
+    env_name = spec['env_name']
+    pipelines_provider = spec['pipelines_provider']
+    provider_name = pipelines_provider['provider']
+
+    error = False
+    if provider_name == 'jenkins':
+        instance_name = pipelines_provider['instance']['name']
+        job_name = get_openshift_saas_deploy_job_name(
+            saas_file_name, env_name, settings)
+        if job_name not in already_triggered:
+            logging.info(['trigger_job', instance_name, job_name])
+            if dry_run:
+                already_triggered.append(job_name)
+
+        if not dry_run:
+            jenkins = jenkins_map[instance_name]
+            try:
+                if job_name not in already_triggered:
+                    jenkins.trigger_job(job_name)
+                    already_triggered.append(job_name)
+                state_update_method(spec)
+            except Exception as e:
+                error = True
```

Comment:
There is a fair bit of duplication here. Maybe:

```python
if job_name not in already_triggered:
    logging.info(['trigger_job', instance_name, job_name])
    if not dry_run:
         jenkins.trigger_job(job_name)
     already_triggered.append(job_name)
```
and let the caller report exceptions.

Reply:
The reason for this Except is that we don't know what may pop up from jenkins. If i could avoid it, i would. I'd hate to let things rise on. As for the set vs list, yes, a set would probably make sense. Do note that this list likely contains a very small number of items (how many jobs are triggered every minute?), so performance is not a concern IMO.

Reply:
We disagree here: do let things rise on and capture them in the `run` method. That's the good thing of exceptions, you capture them where they're easiest to handle. Also, it's less code overall, which also leads to fewer bugs.